### PR TITLE
Improves z level struff

### PR DESCRIPTION
**Pull request by Karolis2011**
- [x] Improves mimic according to lohikar's PR Baystation12/Baystation12#26862
- [x] Adds look up verb
- [x] Adds cross z level talking

### CODE REVIEW

File: code/__defines/_layers.dm [hunk 1]
```diff
@@ -1,8 +1,11 @@
+#define PLANE_SPACE -99
 #define PLANE_SPACE_BACKGROUND -98
 #define PLANE_SPACE_PARALLAX (PLANE_SPACE_BACKGROUND + 1) // -97
 #define PLANE_SPACE_DUST (PLANE_SPACE_PARALLAX + 1) // -96
 #define PLANE_ABOVE_PARALLAX (PLANE_SPACE_BACKGROUND + 3) // -95
 
+// Openspace uses planes -80 through -74
+#define PLANE_OVER_OPENSPACE -3
```
Comment (Lohikar) [inline code/__defines/_layers.dm]: This does not exist on aurora.

File: code/__defines/_layers.dm [hunk 2]
```diff
@@ -1,8 +1,11 @@
+#define PLANE_SPACE -99
```
Comment (Lohikar) [inline code/__defines/_layers.dm]: Space already has a plane, what's this for?

File: code/controllers/subsystems/zcopy.dm [hunk 3]
```diff
@@ -17,78 +17,73 @@
...
 
 /datum/controller/subsystem/zcopy/stat_entry()
-	..("Q:{T:[queued_turfs.len - (qt_idex - 1)]|O:[queued_overlays.len - (qo_idex - 1)]} T:{T:[openspace_turfs.len]|O:[openspace_overlays.len]}")
+	..("Q:{T:[queued_turfs.len - (qt_idex - 1)]|O:[queued_overlays.len - (qo_idex - 1)]} T:{T:[openspace_turfs]|O:[openspace_overlays]}")
 
 /datum/controller/subsystem/zcopy/Initialize(timeofday)
-	starlight_enabled = config.starlight && config.openturf_starlight_permitted
 	// Flush the queue.
 	fire(FALSE, TRUE)
-	if (starlight_enabled)
```
Comment (Lohikar) [inline code/controllers/subsystems/zcopy.dm]: You'll want to keep this block if you want openspace starlight to work.

File: code/controllers/subsystems/zcopy.dm [hunk 4]
```diff
@@ -114,82 +109,91 @@
...
 
-		var/oo_target = OPENTURF_MAX_PLANE - depth
-		var/t_target
+		var/t_target = OPENTURF_MAX_PLANE	// this is where the openturf gets put
 
-		// Handle space parallax & starlight.
-		if (T.is_above_space())
-			t_target = PLANE_SPACE_BACKGROUND
-			if (starlight_enabled && !T.light_range)
-				T.set_light(config.starlight, 0.5)
```
Comment (Lohikar) [inline code/controllers/subsystems/zcopy.dm]: You'll want to keep these two set_light calls if you want openspace starlight.

File: code/controllers/subsystems/zcopy.dm [hunk 5]
```diff
@@ -114,82 +109,91 @@
...
-				// Don't queue deleted stuff or stuff that doesn't need an overlay.
+			if (QDELETED(object) || object.no_z_overlay || object.loc != T.below || object.invisibility == INVISIBILITY_ABSTRACT)
+				// Don't queue deleted stuff, stuff that's not visible, blacklisted stuff, or stuff that's centered on another tile but intersects ours.
 				continue
 
-			if (object.type == /atom/movable/lighting_overlay)	// Special case.
-				T.shadower.copy_lighting(object)
+			// Special case: these are merged into the shadower to reduce memory usage.
+			if (object.type == /atom/movable/lighting_overlay)
+				// T.shadower.copy_lighting(object)
```
Comment (Lohikar) [inline code/controllers/subsystems/zcopy.dm]: The existing aurora copy_lighting implementation works fine, no need to comment it out.

File: code/game/atoms_movable.dm [hunk 6]
```diff
@@ -38,6 +38,9 @@
 			pulledby.pulling = null
 		pulledby = null
 
+	if (bound_overlay)
+		QDEL_NULL(bound_overlay)
```
Comment (Lohikar) [inline code/game/atoms_movable.dm]: This already exists elsewhere, no?

File: code/game/turfs/turf.dm [hunk 7]
```diff
@@ -62,6 +62,14 @@
 	else
 		luminosity = 1
 
+	opaque_counter = opacity
```
Comment (Lohikar) [inline code/game/turfs/turf.dm]: This is a bay lighting var, what's it doing here?

File: code/game/turfs/turf.dm [hunk 8]
```diff
@@ -62,6 +62,14 @@
...
+
+	if (mapload && permit_ao)
+		queue_ao()
```
Comment (Lohikar) [inline code/game/turfs/turf.dm]: Doesn't this already exist elsewhere?

File: code/modules/lighting/lighting_turf.dm [hunk 9]
```diff
@@ -7,8 +7,13 @@
 	var/tmp/list/datum/light_source/affecting_lights       // List of light sources affecting this turf.
 	var/tmp/atom/movable/lighting_overlay/lighting_overlay // Our lighting overlay.
 	var/tmp/list/datum/lighting_corner/corners
+	var/opaque_counter
```
Comment (Lohikar) [inline code/modules/lighting/lighting_turf.dm]: This var belongs to baylights.

File: code/modules/lighting/lighting_turf.dm [hunk 10]
```diff
@@ -7,8 +7,13 @@
 	var/tmp/list/datum/light_source/affecting_lights       // List of light sources affecting this turf.
 	var/tmp/atom/movable/lighting_overlay/lighting_overlay // Our lighting overlay.
 	var/tmp/list/datum/lighting_corner/corners
+	var/opaque_counter
 	var/tmp/has_opaque_atom = FALSE // Not to be confused with opacity, this will be TRUE if there's any opaque atom on the tile.
 
+/turf/New()
```
Comment (Lohikar) [inline code/modules/lighting/lighting_turf.dm]: Don't use New() on turf.

File: code/modules/multiz/zmimic/mimic_common.dm [hunk 11]
```diff
@@ -27,11 +27,20 @@
  */
 /turf/proc/is_above_space()
 	var/turf/T = GetBelow(src)
-	while (T && (T.flags & MIMIC_BELOW))
+	while (T && (T.flags & ZM_MIMIC_BELOW))
 		T = GetBelow(T)
 
 	return istype(T, /turf/space)
 
+/turf/update_icon()
```
Comment (Lohikar) [inline code/modules/multiz/zmimic/mimic_common.dm]: This won't work reliably on aurora.

File: code/modules/multiz/zmimic/mimic_common.dm [hunk 12]
```diff
@@ -27,11 +27,20 @@
...
+	..()
+	if (above)
+		update_above()
+
+/atom/movable/update_icon()
```
Comment (Lohikar) [inline code/modules/multiz/zmimic/mimic_common.dm]: Nor will this.

File: code/modules/multiz/zmimic/mimic_movable.dm [hunk 13]
```diff
@@ -112,33 +122,37 @@
...
 		)
 
-	if (our_overlays || priority_overlays)
-		compile_overlays()
-	else if (bound_overlay)
-		// compile_overlays() calls update_above().
+	var/turf/parent = loc
+	ASSERT(isturf(parent))
+	if (LAZYLEN(parent.ao_overlays_mimic))
+		overlays += parent.ao_overlays_mimic
```
Comment (Lohikar) [inline code/modules/multiz/zmimic/mimic_movable.dm]: Don't manually manipulate overlays, the existing compile_overlays code you replaced was correct.

**Comment by Geevies** [comment]
@Karolis2011 Fixing this?

File: code/__defines/zmimic.dm [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+#define ZM_ALLOW_LIGHTING  8	// If this turf should permit passage of lighting.
+#define ZM_ALLOW_ATMOS    16	// If this turf permits passage of air.
+#define ZM_MIMIC_NO_AO    32	// If the turf shouldn't apply regular turf AO and only do Z-mimic AO.
+
+// Convenience flag.
+#define ZM_MIMIC_DEFAULTS (ZM_MIMIC_BELOW|ZM_ALLOW_LIGHTING)
+
+// For debug purposes, should contain the above defines in ascending order.
+var/list/mimic_defines = list(
+	"NOJAUNT",
```
Comment (skull132) [inline code/__defines/zmimic.dm]: This isn't a define above. :thinking:

File: code/controllers/subsystems/zcopy.dm [hunk 15]
```diff
@@ -17,78 +17,73 @@
...
-		var/turf/T = get_turf(AM)
-		if (TURF_IS_MIMICING(T))
-			if (!(T.flags & MIMIC_QUEUED))
+	log_debug("SSzcopy: update_all() invoked.")
+
+	var/turf/T 	// putting the declaration up here totally speeds it up, right?
+	var/num_upd = 0
+	var/num_del = 0
+	var/num_amupd = 0
+	for (var/atom/A in world)
```
Comment (skull132) [inline code/controllers/subsystems/zcopy.dm]: I wonder what the full initialization time for this is. :thinking:

**Comment by Karolis2011** [comment]
Honestly this is a mess. Might reopen when I will be in the mood.

**Pull request closed without merging**